### DISCUSSION TITLE
Syntax error: Unknown identifier

### DISCUSSION THREAD

**Discussion by LiuPengPeter** [idaholab/moose#16197] [Q&A Modules: General] [answered]
Hello everyone,

I have a problem with my .i file which shows that:

```
Invalid function
1/Vm * ((1-c_Cu)*g_Alpha_Cu + c_Cu*g_Alpha_Sn + R*T*((1-c_Cu)*log(1-c_Cu) + c_Cu*log(c_Cu)) + c_Cu*(1-c_Cu) * (l_Alpha_0 + l_Alpha_1*(1-2*c_Cu)))
c_Cu,Vm,g_Alpha_Cu,g_Alpha_Sn,l_Alpha_0,l_Alpha_1
in ParsedMaterialHelper.
Syntax error: Unknown identifier
```
Does it means that **c_Cu,Vm,g_Alpha_Cu,g_Alpha_Sn,l_Alpha_0,l_Alpha_1** are  unknown?
But I have defined c_Cu in [variable] and others in [Materials] through GenericConstantMaterial.

```
  [./**c_Cu**]
      order = FIRST
      family = LAGRANGE
      initial_condition = 0.02

  [./model_constants]
    type = GenericConstantMaterial
    prop_names = '**Vm** R T'
    prop_values = '16.29e-6 8.314 523'    #Vm:m^3/mol T:K
  [../]
  [./energy_constant]
    type = GenericConstantMaterial
    prop_names = '**g_Alpha_Cu g_Alpha_Sn** g_Ser_Sn g_L_Cu g_L_Sn **l_Alpha_0 l_Alpha_1** l_L_0 l_L_1 l_L_2'
    prop_values = '-19073 -27280 346160 -11083 -28963 -11448 -11694 -10487 -18198 10528.4'    # J/mol
  [../]
```

**Comment by LiuPengPeter** [comment]
Could anyone help? I still couldn't find the bug...
Thank you a lot!

**Reply by aprilnovak** [reply]
I think you have a typo here: `c_Cug_Alpha_Sn`, maybe you are missing a multiplication sign between `c_Cu` and `g_Alpha_Cu`? The parser is interpreting this as a missing symbol for `c_Cug_Alpha_Sn`.

**Reply by LiuPengPeter** [reply]
Sorry. I post it wrong. But that's correct in my .i file.
Could you help me check it?
Thank you a lot!

**Comment by LiuPengPeter** [comment]
My .i file is:
 
```
 [./f_Cu]
    type = ParsedMaterial
    f_name = f_Cu
    args = 'c_Cu'
    material_property_names = 'Vm g_Alpha_Cu g_Alpha_Sn l_Alpha_0 l_Alpha_1'
    # About unit of f: the unit of Vm is m^3/mol. c is the molar percenage of Sn. f is the free energy density.
    function = '1/Vm * ((1-c_Cu)*g_Alpha_Cu + c_Cu*g_Alpha_Sn + R*T*((1-c_Cu)*log(1-c_Cu) + c_Cu*log(c_Cu)) + c_Cu*(1-c_Cu) * (l_Alpha_0 + l_Alpha_1*(1-2*c_Cu)))'
    outputs = exodus
  [../]
```

Could anyone help me check? My brain is full of water now....

**Reply by aprilnovak** [reply]
You need to include multiplication signs in a few places - my previous comment was about two of your variables merging into one long name, but for instance I see you also have a problem with `(1-c_Cu)g_Alpha_Cu ` - you need to change this to `(1-c_Cu)*g_Alpha_Cu `.  There are a few other places this happens, too.

**Reply by LiuPengPeter** [reply]
Thank you for your patience. But there is *. The first post is wrong by mistake. Please see my .i file.
It is:

```
 [./f_Cu]
    type = ParsedMaterial
    f_name = f_Cu
    args = 'c_Cu'
    material_property_names = 'Vm g_Alpha_Cu g_Alpha_Sn l_Alpha_0 l_Alpha_1'
    function = '1/Vm * ((1-c_Cu)*g_Alpha_Cu + c_Cu*g_Alpha_Sn + R*T*((1-c_Cu)*log(1-c_Cu) + c_Cu*log(c_Cu)) + c_Cu*(1-c_Cu) * (l_Alpha_0 + l_Alpha_1*(1-2*c_Cu)))'
    outputs = exodus
  [../]
```

**Reply by dschwen** [reply]
Liu Peng, please use the ``` markup to format excerpts of input files. It is hard to read otherwise.

**Reply by dschwen** [reply]
`R` and `T` are not listed in `args` or `material_property_names`

**Reply by LiuPengPeter** [reply]
Thanks a lot! 
I made such low mistake...

**Comment by dschwen** [comment, marked as answer]
`R` and `T` are not listed in `args` or `material_property_names`
